cadc-storage-adapter-fs/src/main/java/org/opencadc/inventory/storage/fs/FileSystemStorageAdapter.java

```diff
@@ -0,0 +1,523 @@
+                    int colonIndex = bucket.indexOf(":");
+                    path.append(bucket.substring(0, colonIndex + 1));
+                    path.append(File.separator);
+                    path.append(bucket.substring(colonIndex + 1));
+                }
+                if (path.length() > 0) {
+                    fixedParentDir = path.toString();
+                }
+                break;
+            case URI_BUCKET_BASED:
+                if (bucket != null) {
+                    if (bucket.length() > BUCKET_LENGTH) {
+                        throw new IllegalArgumentException("bucket must be a maximum of " + BUCKET_LENGTH + " characters");
+                    }
+                    for (char c : bucket.toCharArray()) {
+                        path.append(c).append(File.separator);
+                    }
+                    bucketDepth = BUCKET_LENGTH - bucket.length();
+                } else {
+                    bucketDepth = BUCKET_LENGTH;
+                }
+                break;
+            default:
+                throw new IllegalStateException("unsupported bucket mode");
+        }
+        try {
+            log.debug("resolving path: " + path.toString());
+            Path bucketPath = root.resolve(path.toString());
+            if (!Files.exists(bucketPath) || !Files.isDirectory(bucketPath)) {
+                throw new IllegalArgumentException("Invalid bucket: " + bucket);
+            }
+            Iterator<StorageMetadata> iterator = null;
+            switch (bucketMode) {
+                case URI_BASED:
+                    iterator = new ArtifactIterator(bucketPath, bucketDepth, fixedParentDir);
+                    break;
+                case URI_BUCKET_BASED:
+                    iterator = new FileSystemIterator(bucketPath, bucketDepth, fixedParentDir);
+                    break;
+                default:
+                    throw new IllegalStateException("unsupported bucket mode");
+            }
+            return iterator;
+        } catch (InvalidPathException e) {
+            throw new IllegalArgumentException("Invalid bucket: " + bucket);
+        }
+    }
+    
+    private Path createArtifactPath(URI artifactURI, String uriBucket) {
+        StringBuilder path = new StringBuilder();
+        switch (bucketMode) {
+            case URI_BASED:
+                path.append(artifactURI.getScheme());
+                path.append(":").append(File.separator);
+                path.append(artifactURI.getSchemeSpecificPart());
+                break;
+            case URI_BUCKET_BASED:
+                for (char c : uriBucket.toCharArray()) {
+                    path.append(c).append(File.separator);
+                }
+                UUID filename = UUID.randomUUID();
+                path.append(filename);
+                break;
+            default:
+                throw new IllegalStateException("unsupported bucket mode");
+        }
+        return root.resolve(path.toString());
```

Comment:
I wonder if this method should have created and returned a StorageLocation? The put(...) is parsing the path returned here to create the StorageLocation.  You need StorageLocation-to-Path code for get(...) anyway so put() could use that to figure out where to write... 

It seems like all the separate bits are here and they get serialised and re-parsed again.

Reply:
Yep, see reply above.